libtenzir/include/tenzir/concept/parseable/tenzir/data.hpp

```diff
@@ -98,6 +100,16 @@ struct data_parser : parser_base<data_parser> {
     // clang-format on
     return p;
   }
+
+  static bool check_list_types(const std::vector<data>& xs) {
+    if (xs.empty())
+      return true;
+    // TODO: Comparing indices won't catch mismatched nested lists or objects
+    auto first_elem_type = xs.front().get_data().index();
+    return std::all_of(xs.begin() + 1, xs.end(), [&](const auto& elem) {
+      return first_elem_type == elem.get_data().index();
+    });
```

Comment:
Another option would be to defer this until we have generalized expressions a bit further. Currently, expressions are just a set of predicates with very strict rules. However, we want to relax that so that `[foo(bar, baz), 3 + 1, -20.0]` can also be resolved accordingly. That's a bigger effort that we were looking to schedule closer to the end of this year.

Reply:
I ended up reverting this parsing logic change, and doing the changes in `type::infer`. It now does deep type checking for recursive data structures (like lists, maps, and records).